### PR TITLE
include innoextract

### LINKED ISSUE #3578: Include innoextract in launcher
**Describe your proposal**
We should think about including innoextract directly into the source.

Following benefits:
- We could add a button in the launcher to extract gog files for easiest install
- This would work platform independent without extra tools (also mobile installation possible)
- Could be a solution for adding chronicles without copyright problems

I think most users use the gog packages as base.

**Additional context**
I'm not sure how easy or hard this task is. Seems we don't need lzma libs for heroes 3 installer (tested to extract with lzma disabled build).

I'm not able to do this because I know not much enough about c++ building systems.

### PR DESCRIPTION
**Pull request by Laserlicht**
I've tried to include innoextract directly in vcmi.

Basically it works. Only change to upstream innoextract is the `cmakelists.txt`.

Locally on Linux with GCC it works great.
grafik

Sadly there is still a problem with clang: https://github.com/Laserlicht/vcmi/actions/runs/8511235163/job/23310781125
And I think there are some hacky things in `cmakelists.txt`-changes.

Do we have some people to help out with the cmake stuff? If you don't have much experience in C++ stuff: its absolutely an nightmare.

fixes #3578

### CODE REVIEW

**Comment by IvanSavenko** [comment]
Many thanks for attempt. Would be great addition to 1.5 if we can implement it, especially for mobile platforms

Perhaps instead of embedding innoextract into vcmi repo we should:
- create a fork of innoextract in vcmi org (I can create it or you can use personal repo for test)
- modify CMakeLists as needed in fork
- add vcmi/innoextract repo as git submodule into vcmi/vcmi

This way we'll be able to clearly see all our changes to innoextract and easily upgrade our innoextract version when needed

**Comment by Laserlicht** [comment]
> This way we'll be able to clearly see all our changes to innoextract and easily upgrade our innoextract version when needed

Yes. I had the same idea in mind. But my first goal was to bring it to work.

> Would be great addition to 1.5 if we can implement it, especially for mobile platforms

Would be great and would solve many problems caused by [Layer 8](https://www.ionos.com/digitalguide/server/security/layer-8-problem/). I think for this we would need a second QFileDialog for the bin file and copy both to a temporary folder to avoid sandbox problems.

File: launcher/ext/innoextract/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+	include(GNUInstallDirs)
+endif()
+
+
+# Helper scrips
+
+include(CheckSymbolExists)
+
+set(CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}/cmake") # For custom cmake modules
+include(BuildType)
```
Comment (IvanSavenko) [inline launcher/ext/innoextract/CMakeLists.txt:109]: This include can make quite a lot of changes to build config.
That clang linkage error might come from here.

Comment (Laserlicht) [reply launcher/ext/innoextract/CMakeLists.txt]: Nice, thanks! Locally it's build now. Let's look what the CI do.

**Comment by IvanSavenko** [comment]
MacOS CI failure - looks like innoextract or vcmilauncher was no linked to boost_iostreams.
Explicitly linking iostreams to vcmilauncher might help. Not sure how boost linking works on macOS/conan, @kambala-decapitator might know more.

**Comment by Laserlicht** [comment]
Yes, I've added iostreams on several places (VCMI wasn't using this part of boost yet).
E.g. in "main" `CMakeLists.txt`.

But I think I've found it... Let's have a look...

**Comment by Alexander-Wilms** [comment]
You could temporarily add `fail-fast: false` to .github/workflows/github.yml so you don't have to fix the builds one after the other:

https://stackoverflow.com/a/65647509/2278742

**Comment by Laserlicht** [comment]
> wouldn't innoextract be possible to get added as git submodule?

Yes. If everything is working. :)

**Comment by kambala-decapitator** [comment]
boost needs to be rebuilt on the Conan platforms

**Review by kambala-decapitator — COMMENTED** [review]
note that this solution wouldn't work on mobile platforms out of the box because currently there's no ability to select arbitrary directory

**Comment by kambala-decapitator** [comment]
also Conan has innoextract: https://conan.io/center/recipes/innoextract?version=1.9.0

**Comment by IvanSavenko** [comment]
> also Conan has innoextract: https://conan.io/center/recipes/innoextract?version=1.9.0

As a separate executable from what I see. But we mostly need innoextract as part of our single binary for Android / IPhone.

**Comment by Laserlicht** [comment]
> boost needs to be rebuilt on the Conan platforms

Yes. But how?

**Comment by kambala-decapitator** [comment]
> > also Conan has innoextract: https://conan.io/center/recipes/innoextract?version=1.9.0
> 
> As a separate executable from what I see. But we mostly need innoextract as part of our single binary for Android / IPhone.

yeah, indeed, whatever upstream provides

**Comment by kambala-decapitator** [comment]
> > boost needs to be rebuilt on the Conan platforms
> 
> Yes. But how?

currently possible only manually, can do that. So the target is to include this into 1.5?

**Comment by Laserlicht** [comment]
> note that this solution wouldn't work on mobile platforms out of the box because currently there's no ability to select arbitrary directory

Yes the workaround idea would be two file selectors (first exe then bin) and copy it to temp folder (which should be openable?)

File: launcher/firstLaunch/firstlaunch_moc.cpp [hunk 2]
```diff
@@ -21,6 +21,8 @@
 #include "../../lib/filesystem/Filesystem.h"
 #include "../languages.h"
 
+#include "cli/extract.hpp"
```
[the remark this answers is not in the document]
Comment (Laserlicht) [reply launcher/firstLaunch/firstlaunch_moc.cpp:24]: What's the right command to do this on cmake?

**Comment by Laserlicht** [comment]
> > > boost needs to be rebuilt on the Conan platforms
> > 
> > 
> > Yes. But how?
> 
> currently possible only manually, can do that. So the target is to include this into 1.5?

If we get everything working (at least on desktop) until then, then: yes

**Comment by kambala-decapitator** [comment]
btw I think that Windows doesn't need this functionality

**Comment by Laserlicht** [comment]
> btw I think that Windows doesn't need this functionality

You mean the copying of the exe & bin? Yes - I think Windows & Linux doesn't need this. With MacOS I'm not sure.
What's to best way to difference the Platforms? Preprocessor #ifdef?

The two FileSelectors would be also good on these platforms for error detection (and the user knows that he needs two files).

File: launcher/firstLaunch/firstlaunch_moc.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Laserlicht) [reply launcher/firstLaunch/firstlaunch_moc.cpp:24]: @kambala-decapitator How can I fix that. I want to avoid changes in innoextract code itself if it isn't necessary.
Can this be done with cmake?

[the remark this answers is not in the document]
Comment (kambala-decapitator) [reply launcher/firstLaunch/firstlaunch_moc.cpp:24]: > Wrapping only the include isn't working.

too bad, I expected it'd work. I don't think you can solve it purely with cmake. Let's leave as is then.

you could also create a thin wrapper, but not insisting on that.

**Comment by kambala-decapitator** [comment]
> > btw I think that Windows doesn't need this functionality
> 
> You mean the copying of the exe & bin? Yes - I think Windows & Linux doesn't need this. With MacOS I'm not sure. What's to best way to difference the Platforms? Preprocessor #ifdef?
> 
> The two FileSelectors would be also good on these platforms for error detection (and the user knows that he needs two files).

I mean Windows doesn't need innoextract at all as you can install H3 natively. The condition whether to activate innoextract code should be in cmake, something like

```cmake
if(NOT WINDOWS)
  set(USE_INNOEXTRACT ON)
  target_compile_definitions(launcher USE_INNOEXTRACT=1)
endif()
```

(wrote without looking at docs or project, can have mistakes)

**Comment by Laserlicht** [comment]
> I mean Windows doesn't need innoextract at all as you can install H3 natively. The condition whether to activate innoextract code should be in cmake, something like

Ah. Yes Windows don't need it. But even for windows it's quite a nice feature (no need for installing H3) and I would like to keep it.

I placed everything behind a cmake switch. It's now possible to completely disable it (e.g. for specific platforms in CI).

File: CMakeLists.txt [hunk 3]
```diff
@@ -88,6 +88,12 @@ if(ENABLE_ERM AND NOT ENABLE_LUA)
 	set(ENABLE_LUA ON)
 endif()
 
+if(ENABLE_LAUNCHER)
+	option(ENABLE_INNOEXTRACT "Enable innoextract for GOG file extraction in launcher" ON)
+else()
+	set(ENABLE_INNOEXTRACT OFF)
+endif()
+
```
Comment (IvanSavenko) [inline CMakeLists.txt:96]: Not sure if CMake provides any better way to handle such cases, but this might be a little inconvenient:
- disable launcher
- run cmake, do stuff
- re-enable launcher, run cmake

You now have innoextract silently disabled even though this was not your intention and you never touched that option

Feel free to ignore if you can't think of better option(s)

Comment (kambala-decapitator) [reply CMakeLists.txt:96]: maybe dependent option would help https://cmake.org/cmake/help/latest/module/CMakeDependentOption.html#module:CMakeDependentOption

**Comment by kambala-decapitator** [comment]
> How complicated is it to rebuild conan boost? Would like to test the other platforms than Win & Linux.

run command and wait :) Do you have access to macOS / iOS?

**Comment by Laserlicht** [comment]
@kambala-decapitator Can you do that or provide information how to get the CI working? :)

**Comment by kambala-decapitator** [comment]
> @kambala-decapitator Can you do that or provide information how to get the CI working? :)

Can do later.

If you wish to try on your own: simply follow our Conan docs, but pass --build missing instead of --build never

**Comment by Laserlicht** [comment]
> > @kambala-decapitator Can you do that or provide information how to get the CI working? :)
> 
> Can do later.
> 
> If you wish to try on your own: simply follow our Conan docs, but pass --build missing instead of --build never
> 
> After building locally, upload to our Dependencies repo and change URL in the CI script

Would be great if you could do this. :)

I don't have write access to this repositories and no experience in building MacOS stuff.

**Comment by kambala-decapitator** [comment]
@Laserlicht uploaded updated Boost for macOS Intel as depends v1.2.1, please give it a try. I recommend disabling other CI jobs for the time being. Adjust the version here: https://github.com/vcmi/vcmi/blob/develop/CI/mac/before_install.sh#L8

**Comment by kambala-decapitator** [comment]
fixed that error and started CI again

**Comment by Laserlicht** [comment]
@kambala-decapitator Thanks! Works on MacOS. Can you also build iOS? :)

**Comment by kambala-decapitator** [comment]
uploaded iOS 1.2.1: https://github.com/vcmi/vcmi-ios-deps/releases/tag/1.2.1

**Comment by Laserlicht** [comment]
Nice! I've installed VCMI only with IPA, and GOG-files (exe+bin).

Works without problems! Thanks @kambala-decapitator

Now we need only to update MacOS ARM and the two Android's. @kambala-decapitator Is it possible to do that also for me? Then I can finish the PR.

**Comment by IvanSavenko** [comment]
Two Androids would have to wait for kambala's launcher PR.
At the moment Android still uses own java launcher

**Comment by rilian-la-te** [comment]
Can innoextract be optional for "normal" linux?

**Comment by Laserlicht** [comment]
> Can innoextract be optional for "normal" linux?

There is a `ENABLE_INNOEXTRACT` option. If disabled it will completely build without innoextract. But default is `ON`.

**Comment by Laserlicht** [comment]
> Re-enable Android CI but disable innoextract for it. It actually should be disabled already - since Androids have `ENABLE_LAUNCHER` set to OFF, so innoextract option should also be off

Fails in conan setup :(

**Comment by kambala-decapitator** [comment]
> Fails in conan setup :(

either wait for the launcher PR or adjust conanfile.py to use new boost params on non-Android:

    self.options["boost"].without_iostreams = self.settings.os == "Android"

**Comment by kambala-decapitator** [comment]
uploaded macOS ARM

**Comment by Laserlicht** [comment]
> uploaded macOS ARM

Good work! Thanks a lot. :)

File: docs/players/Installation_iOS.md [hunk 4]
```diff
@@ -24,17 +24,21 @@ To install the signed ipa on your device, you can use Xcode or Apple Configurato
 
 Note: if you don't need in-game videos, you can omit downloading/copying file VIDEO.VID from data folder - it will save your time and space. The same applies to the Mp3 directory.
 
-### Step 2.a: Installing data files with Finder or Windows explorer
+### Step 2.a: Installing data files with GOG offline installer
+
+If you buyed HoMM3 on [GOG](https://www.gog.com/de/game/heroes_of_might_and_magic_3_complete_edition) you can download the files directly with the browser and install it in launcher. You need to select the exe file first and afterwards the bin file. This process takes a few seconds. Be patient.
```
Comment (IvanSavenko) [inline docs/players/Installation_iOS.md:29]:
```suggestion
If you bought HoMM3 on [GOG](https://www.gog.com/de/game/heroes_of_might_and_magic_3_complete_edition), you can download the files directly from the browser and install them in the launcher. Select the .exe file first, then the .bin file. This may take a few seconds. Please be patient.
```

File: docs/players/Installation_macOS.md [hunk 5]
```diff
@@ -15,5 +15,11 @@ Please report about gameplay problem on forums: [Help & Bugs](https://forum.vcmi
 
 # Step 2: Installing Heroes III data files
 
+### Step 2.a: Installing data files with GOG offline installer
+
+If you buyed HoMM3 on [GOG](https://www.gog.com/de/game/heroes_of_might_and_magic_3_complete_edition) you can download the files directly with the browser and install it in launcher. You need to select the exe file first and afterwards the bin file. This process takes a few seconds. Be patient.
```
Comment (IvanSavenko) [inline docs/players/Installation_macOS.md:20]:
```suggestion
If you bought HoMM3 on [GOG](https://www.gog.com/de/game/heroes_of_might_and_magic_3_complete_edition), you can download the files directly from the browser and install them in the launcher. Select the .exe file first, then the .bin file. This may take a few seconds. Please be patient.
```

File: launcher/firstLaunch/firstlaunch_moc.cpp [hunk 6]
```diff
@@ -279,7 +293,69 @@ QString FirstLaunchView::getHeroesInstallDir()
...
+	QString filterBin = tr("GOG bin file") + " (*.bin)";
+	QString fileBin = QFileDialog::getOpenFileName(this, tr("Select a GOG data (bin) file..."), QFileInfo(fileExe).absolutePath(), filterBin);
+	if(fileBin.isEmpty())
+		return;
+
+	if(QFileInfo(fileExe).completeBaseName() + QString("-1") != QFileInfo(fileBin).completeBaseName())
+	{
+		QMessageBox::critical(this, tr("File name error!"), tr("The filename of exe and bin has to be compatible!"), QMessageBox::Ok, QMessageBox::Ok);
+		return;
+	}
```
Comment (IvanSavenko) [inline launcher/firstLaunch/firstlaunch_moc.cpp:313]: Perhaps we can modify innoextract and provide both .bin and .exe paths separately into innoextract, even if they don't match? Or too much of a trouble?

If not - error message needs to be more clear. What is 'compatible'?

File: launcher/firstLaunch/firstlaunch_moc.cpp [hunk 7]
```diff
@@ -279,7 +293,69 @@ QString FirstLaunchView::getHeroesInstallDir()
...
+			process_file(tmpFileExe.toStdString(), o);
+
+			QStringList dirData = QDir(dir.path()).entryList({"data"}, QDir::Filter::Dirs);
+			if(dirData.empty() || QDir(QDir(dir.path()).filePath(dirData.front())).entryList({"*.lod"}, QDir::Filter::Files).empty())
+			{
+				QMessageBox::critical(this, tr("No Heroes III data!"), tr("Selected files does not contain Heroes III data!"), QMessageBox::Ok, QMessageBox::Ok);
+				ui->pushButtonGogInstall->setText(tr("Install GOG files"));
+				return;
+			}
+			copyHeroesData(dir.path(), true);
```
Comment (IvanSavenko) [inline launcher/firstLaunch/firstlaunch_moc.cpp:352]: We might need to clean up afterwards - delete all extracted files other than data/maps/mp3

File: launcher/firstLaunch/firstlaunch_moc.cpp [hunk 8]
```diff
@@ -279,7 +293,69 @@ QString FirstLaunchView::getHeroesInstallDir()
...
+			o.output_dir = dir.path().toStdString();
+			o.extract_temp = true;
+			o.extract_unknown = true;
+			o.filenames.set_expand(true);
+
+			o.preserve_file_times = true; // also correctly closes file -> without it: on Windows the files are not written completly
+
+			process_file(tmpFileExe.toStdString(), o);
+
+			QStringList dirData = QDir(dir.path()).entryList({"data"}, QDir::Filter::Dirs);
```
Comment (IvanSavenko) [inline launcher/firstLaunch/firstlaunch_moc.cpp:345]: I think our current file copy algorithm actually looks up directories to copy in case-insensitive way. Perhaps use same approach here?

File: launcher/firstLaunch/firstlaunch_moc.cpp [hunk 9]
```diff
@@ -279,7 +293,69 @@ QString FirstLaunchView::getHeroesInstallDir()
...
+	if(QFileInfo(fileExe).completeBaseName() + QString("-1") != QFileInfo(fileBin).completeBaseName())
+	{
+		QMessageBox::critical(this, tr("File name error!"), tr("The filename of exe and bin has to be compatible!"), QMessageBox::Ok, QMessageBox::Ok);
+		return;
+	}
+
+	QTimer::singleShot(100, this, [this, fileExe, fileBin](){ // background to make sure FileDialog is closed...
+		QTemporaryDir dir;
+		if(dir.isValid() && QFile::exists(fileExe) && QFile::exists(fileBin)) {
+			ui->pushButtonGogInstall->setText(tr("Installing... Please wait!"));
```
Comment (IvanSavenko) [inline launcher/firstLaunch/firstlaunch_moc.cpp:318]: Perhaps also block button while process is ongoing? Or even block whole UI if Qt allows that.

File: launcher/firstLaunch/firstlaunch_moc.cpp (hunk 7 above)
Comment (Laserlicht) [reply launcher/firstLaunch/firstlaunch_moc.cpp:352]: Not needed. Temp folder cleans up on destruction of object: https://doc.qt.io/qt-6/qtemporarydir.html

File: launcher/firstLaunch/firstlaunch_moc.cpp (hunk 9 above)
Comment (Laserlicht) [reply launcher/firstLaunch/firstlaunch_moc.cpp:318]: Yes the UI is already completely blocked. Not the nicest way - async would be great (as the process needs some seconds). But I think okay for now.

File: launcher/firstLaunch/firstlaunch_moc.cpp (hunk 8 above)
Comment (Laserlicht) [reply launcher/firstLaunch/firstlaunch_moc.cpp:345]: I've copied the approach from here: https://github.com/vcmi/vcmi/blob/de9dbf0cd78c300dcd1663fc41615b52aadcf0b4/launcher/firstLaunch/firstlaunch_moc.cpp#L380

Basically only to check if we have some H3-like output from innoextract (as you can also feed it with other games for example).

File: launcher/firstLaunch/firstlaunch_moc.cpp (hunk 7 above)
Comment (IvanSavenko) [reply launcher/firstLaunch/firstlaunch_moc.cpp:352]: Potential problem is that this temporary directory might be created on a different partition. Meaning:
- free space might be limited
- we won't be able to quickly move files from temporary directory to data directory if it is on different partition

Or it might not be a problem. Other than Linux where /tmp and /home are kind of recommended to be placed on different partitions, on rest of OS's our data directory is usually on same partition as OS and temp directories.

Comment (Laserlicht) [reply launcher/firstLaunch/firstlaunch_moc.cpp:346]: Yes. Maybe not the perfect solution. But even under Linux it's the usual way to have the temp folder under the main partition. At least for the standard settings of many distributions-installer.

But I would leave it as it is. And maybe change later.

Comment (IvanSavenko) [reply launcher/firstLaunch/firstlaunch_moc.cpp:346]: While temp folder aka `/tmp` is usually on main partition, `/home` including /home/me/.local/share/vcmi (which is where we save h3 data) may be on a different partition.

But yeah, let's keep as it for now. Unlikely to be an issue. 1 Gb of temp files is low amount nowadays.

**Review by IvanSavenko — APPROVED** [review]

**Comment by kambala-decapitator** [comment]
Will review today or tomorrow

File: launcher/firstLaunch/firstlaunch_moc.cpp [hunk 10]
```diff
@@ -279,7 +293,63 @@ QString FirstLaunchView::getHeroesInstallDir()
 	return QString{};
 }
 
-void FirstLaunchView::copyHeroesData(const QString & path)
+void FirstLaunchView::extractGogData()
+{
+#ifndef VCMI_NO_INNOEXTRACT
+	QString filterExe = tr("GOG executable") + " (*.exe)";
+	QString fileExe = QFileDialog::getOpenFileName(this, tr("Select a GOG installer (exe) file..."), QDir::homePath(), filterExe);
```
Comment (kambala-decapitator) [inline launcher/firstLaunch/firstlaunch_moc.cpp:300]: So this works on iOS? Nice!

File: launcher/firstLaunch/firstlaunch_moc.cpp [hunk 11]
```diff
@@ -279,7 +293,63 @@ QString FirstLaunchView::getHeroesInstallDir()
...
+	if(fileExe.isEmpty())
+		return;
+
+	QString filterBin = tr("GOG bin file") + " (*.bin)";
+	QString fileBin = QFileDialog::getOpenFileName(this, tr("Select a GOG data (bin) file..."), QFileInfo(fileExe).absolutePath(), filterBin);
```
Comment (kambala-decapitator) [inline launcher/firstLaunch/firstlaunch_moc.cpp:305]: You could check automatically if bin file with the same name exists and suggest to select it only if not found. But not sure if this would work on mobile platforms.

File: launcher/firstLaunch/firstlaunch_moc.cpp [hunk 12]
```diff
@@ -279,7 +293,63 @@ QString FirstLaunchView::getHeroesInstallDir()
...
+	if(fileBin.isEmpty())
+		return;
+
+	QTimer::singleShot(100, this, [this, fileExe, fileBin](){ // background to make sure FileDialog is closed...
+		QTemporaryDir dir;
+		if(dir.isValid() && QFile::exists(fileExe) && QFile::exists(fileBin)) {
```
Comment (kambala-decapitator) [inline launcher/firstLaunch/firstlaunch_moc.cpp:311]: Not sure what's the point of checking for existence of files that use has just selected

File: launcher/firstLaunch/firstlaunch_moc.cpp [hunk 13]
```diff
@@ -279,7 +293,63 @@ QString FirstLaunchView::getHeroesInstallDir()
...
+			QString tmpFileExe = dir.filePath("h3_gog.exe");
+			QFile(fileExe).copy(tmpFileExe);
+			QFile(fileBin).copy(dir.filePath("h3_gog-1.bin"));
+
+			::extract_options o;
+			o.extract = true;
+
+			// standard settings
+			o.gog_galaxy = true;
+			o.codepage = (uint32_t)0;
```
Comment (kambala-decapitator) [inline launcher/firstLaunch/firstlaunch_moc.cpp:329]:
```suggestion
			o.codepage = 0U;
```

File: launcher/firstLaunch/firstlaunch_moc.cpp [hunk 14]
```diff
@@ -279,7 +293,63 @@ QString FirstLaunchView::getHeroesInstallDir()
...
+			o.extract_temp = true;
+			o.extract_unknown = true;
+			o.filenames.set_expand(true);
+
+			o.preserve_file_times = true; // also correctly closes file -> without it: on Windows the files are not written completly
+
+			process_file(tmpFileExe.toStdString(), o);
+
+			QStringList dirData = QDir(dir.path()).entryList({"data"}, QDir::Filter::Dirs);
+			if(dirData.empty() || QDir(QDir(dir.path()).filePath(dirData.front())).entryList({"*.lod"}, QDir::Filter::Files).empty())
```
Comment (kambala-decapitator) [inline launcher/firstLaunch/firstlaunch_moc.cpp:340]: Please create the QDir object only once

File: launcher/firstLaunch/firstlaunch_moc.cpp [hunk 15]
```diff
@@ -279,7 +293,63 @@ QString FirstLaunchView::getHeroesInstallDir()
...
+			{
+				QMessageBox::critical(this, tr("No Heroes III data!"), tr("Selected files does not contain Heroes III data!"), QMessageBox::Ok, QMessageBox::Ok);
```
Comment (kambala-decapitator) [inline launcher/firstLaunch/firstlaunch_moc.cpp:342]:
```suggestion
				QMessageBox::critical(this, tr("No Heroes III data!"), tr("Selected files do not contain Heroes III data!"), QMessageBox::Ok, QMessageBox::Ok);
```

File: CMakeLists.txt [hunk 16]
```diff
@@ -190,6 +193,10 @@ else()
 	endmacro(enable_pch)
 endif()
 
+if(NOT ENABLE_INNOEXTRACT)
+	add_definitions(-DVCMI_NO_INNOEXTRACT)
```
Comment (kambala-decapitator) [inline CMakeLists.txt:197]: Why not calling it VCMI_INNOEXTRACT? (Without negation)

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]

File: launcher/firstLaunch/firstlaunch_moc.cpp [hunk 17]
```diff
@@ -279,7 +293,63 @@ QString FirstLaunchView::getHeroesInstallDir()
...
+	QString fileExe = QFileDialog::getOpenFileName(this, tr("Select a GOG installer (exe) file..."), QDir::homePath(), filterExe);
+	if(fileExe.isEmpty())
+		return;
+
+	QString filterBin = tr("GOG bin file") + " (*.bin)";
+	QString fileBin = QFileDialog::getOpenFileName(this, tr("Select a GOG data (bin) file..."), QFileInfo(fileExe).absolutePath(), filterBin);
+	if(fileBin.isEmpty())
+		return;
+
+	QTimer::singleShot(100, this, [this, fileExe, fileBin](){ // background to make sure FileDialog is closed...
```
Comment (kambala-decapitator) [inline launcher/firstLaunch/firstlaunch_moc.cpp:309]: Why is this needed?

File: CMakeLists.txt (hunk 16 above)
Comment (IvanSavenko) [reply CMakeLists.txt:197]: Editor and launcher use -DENABLE_EDITOR / -DENABLE_EDITOR definition. So by same logic this one should be named -DENABLE_INNOEXTRACT

File: launcher/firstLaunch/firstlaunch_moc.cpp (hunk 10 above)
Comment (Laserlicht) [reply launcher/firstLaunch/firstlaunch_moc.cpp:300]: Yes, works without problem. :)
Only not so nice thing is that the Filter doesn't work on iOS (I think it's not supported by iOS).

File: launcher/firstLaunch/firstlaunch_moc.cpp (hunk 11 above)
Comment (Laserlicht) [reply launcher/firstLaunch/firstlaunch_moc.cpp:305]: Yes. But I think I would wait for Android and maybe simplify it later, if possible.

File: launcher/firstLaunch/firstlaunch_moc.cpp (hunk 17 above)
Comment (Laserlicht) [reply launcher/firstLaunch/firstlaunch_moc.cpp:309]: It's a kind of workaround. The unpacking process blocks the UI.

Without it: On my testing system it doesn't even close the FileSelector until the process is done. I've tried to force update the UI before extracting, but it doesn't work. :(

With this workaround everything works great. The FileSelector is closed, the button changed name and color.

File: CMakeLists.txt (hunk 16 above)
Comment (Laserlicht) [reply CMakeLists.txt:197]: Yes, that's the reason I preferred this.

**Comment by Laserlicht** [comment]
Thanks @kambala-decapitator

File: launcher/firstLaunch/firstlaunch_moc.cpp (hunk 17 above)
Comment (kambala-decapitator) [reply launcher/firstLaunch/firstlaunch_moc.cpp:309]: > I've tried to force update the UI 

So calling qApp->processEvents() right before starting extraction doesn't help?

Anyway, extraction should be done on background thread (use QtConcurrent) and ui should show a spinner or indefinite progress bar (if the library doesn't provide progress updates).

File: CMakeLists.txt (hunk 16 above)
Comment (kambala-decapitator) [reply CMakeLists.txt:197]: Sorry, I don't get the reason :) 

Also, the define should be added only to the launcher target.

File: launcher/firstLaunch/firstlaunch_moc.cpp [hunk 18]
```diff
@@ -279,7 +293,65 @@ QString FirstLaunchView::getHeroesInstallDir()
...
+		QTemporaryDir dir;
+		if(dir.isValid()) {
+			QDir tempDir = QDir(dir.path());
```
Comment (kambala-decapitator) [inline launcher/firstLaunch/firstlaunch_moc.cpp:312]: If you will be pushing another commit, let's also improve this

```suggestion
			QDir tempDir{dir.path()};
```

File: launcher/firstLaunch/firstlaunch_moc.cpp (hunk 17 above)
Comment (Laserlicht) [reply launcher/firstLaunch/firstlaunch_moc.cpp:309]: No. Tested it and it wouldn't help. :(

Yes. Would also prefer this. But my knowledge to QT is very limited now. The first project with it. So I have no idea how to do it the clean way.

File: CMakeLists.txt (hunk 16 above)
Comment (Laserlicht) [reply CMakeLists.txt:197]: Ah, sorry. Misunderstood it.

The reason for this way was `VCMI_NO_EXTRA_VERSION`. Used this as reference.

**Review by kambala-decapitator — APPROVED** [review]

**Comment by kambala-decapitator** [comment]
When can it be merged?
